rust/src/tx_builder.rs

```diff
@@ -265,118 +271,252 @@ impl TransactionBuilder {
     pub fn add_inputs_from(&mut self, inputs: &TransactionUnspentOutputs, strategy: CoinSelectionStrategyCIP2) -> Result<(), JsError> {
-        let mut available_inputs = inputs.0.clone();
+        let available_inputs = &inputs.0.clone();
         let mut input_total = self.get_total_input()?;
         let mut output_total = self
             .get_explicit_output()?
             .checked_add(&Value::new(&self.get_deposit()?))?
             .checked_add(&Value::new(&self.min_fee()?))?;
         match strategy {
             CoinSelectionStrategyCIP2::LargestFirst => {
-                available_inputs.sort_by_key(|input| input.output.amount.coin);
-                // iterate in decreasing order of ADA-only value
-                for input in available_inputs.iter().rev() {
-                    if input_total >= output_total {
-                        break;
+                if self.outputs.0.iter().any(|output| output.amount.multiasset.is_some()) {
+                    return Err(JsError::from_str("Multiasset values not supported by LargestFirst. Please use LargestFirstMultiAsset"));
+                }
+                self.cip2_largest_first_by(
+                    available_inputs,
+                    &mut (0..available_inputs.len()).collect(),
+                    &mut input_total,
+                    &mut output_total,
+                    |value| Some(value.coin))?;
+            },
+            CoinSelectionStrategyCIP2::RandomImprove => {
+                if self.outputs.0.iter().any(|output| output.amount.multiasset.is_some()) {
+                    return Err(JsError::from_str("Multiasset values not supported by RandomImprove. Please use RandomImproveMultiAsset"));
+                }
+                use rand::Rng;
+                let mut rng = rand::thread_rng();
+                let mut available_indices = (0..available_inputs.len()).collect::<BTreeSet<usize>>();
+                self.cip2_random_improve_by(
+                    available_inputs,
+                    &mut available_indices,
+                    &mut input_total,
+                    &mut output_total,
+                    |value| Some(value.coin),
+                    &mut rng)?;
+                // Phase 3: add extra inputs needed for fees (not covered by CIP-2)
+                // We do this at the end because this new inputs won't be associated with
+                // a specific output, so the improvement algorithm we do above does not apply here.
+                while input_total.coin < output_total.coin {
+                    if available_indices.is_empty() {
+                        return Err(JsError::from_str("UTxO Balance Insufficient[x]"));
                     }
-                    // differing from CIP2, we include the needed fees in the targets instead of just output values
+                    let i = *available_indices.iter().nth(rng.gen_range(0..available_indices.len())).unwrap();
+                    available_indices.remove(&i);
```

Comment:
Using a Vec instead of BTreeSet is better for performance here. Because you select a random element and selecting random elements from a Vec is constant time. You can move the last element to the removed element's place when removing an element if you use a Vec.

Reply:
You have to look at where it's used within `random_improve_by()` too where it's passed into. One thing to remember there is that you're indexing it by value not by index so you'd have an O(n) look-up (e.g. for `position()` even if the `swap_remove()` call itself is O(1). You can take out the parameter to `nth()` call here in Phase 3 but in Phase 1/2 we can't re-use the `random_index` down there (in `random_improve_by()`) either to pass into `swap_remove()` because that's an index into `relevant_indices` not `available_indices` so the indices won't have anything in common - but the values inside them will. I actually had `available_indices` as a Vec like in largest-first before changing it. Phase 3 is also likely to be run very few times relative to Phase 1/2 as it is only to cover the difference in fees caused by the inputs added in phases 1/2. The bulk of the ADA should be added beforehand. You're choosing between O(n) vs O(logn) possibly a bunch of times in phase 1/2 followed by O(1) vs O(n) maybe once or twice in phase 3.

There are, however, differences in cache/locality/allocations although BTreeMap has contiguous sections within it so I'm not sure if it is even much of an issue in practice vs Vec. I have no idea how all this translates into wasm though which is where most of our users are using it from. It's possible this could be significant relative to algorithmic differences but until we start to see it being an issue I'm not sure how much we should care.

Reply:
Thanks for the explanation. I agree this should be good than